File: apps/desktop/src/lib/api/logs.ts
import type { LogsResponse, RefreshLog, ScriptRunLog } from "../../types/core";
import { requestJson } from "./client";

export async function fetchRefreshLogs(
  options?: {
    limit?: number;
    offset?: number;
    status?: "running" | "success" | "failed";
    sourceId?: string;
    includeScriptLogs?: boolean;
  },
): Promise<LogsResponse> {
  const limit = options?.limit ?? 20;
  const offset = options?.offset ?? 0;
  const status = options?.status;
  const sourceId = options?.sourceId;
  const includeScriptLogs = options?.includeScriptLogs ?? false;

  const params = new URLSearchParams();
  params.set("limit", String(limit));
  params.set("offset", String(offset));
  if (status) {
    params.set("status", status);
  }
  if (sourceId) {
    params.set("source_id", sourceId);
  }
  if (includeScriptLogs) {
    params.set("include_script_logs", "true");
  }

  const payload = await requestJson<{
    logs: Array<{
      id: string;
      source_id: string;
      source_name?: string | null;
      trigger_type: string;
      status: string;
      started_at?: string | null;
      finished_at?: string | null;
      node_count?: number | null;
      error_code?: string | null;
      error_message?: string | null;
      script_logs?: Array<{
        id: string;
        source_id: string;
        plugin_id: string;
        level: string;
        message: string;
        created_at: string;
      }>;
    }>;
    pagination?: {
      limit: number;
      offset: number;
      total: number;
      has_more: boolean;
    };
  }>("GET", `/api/logs?${params.toString()}`);

  const logs: RefreshLog[] = payload.logs.map((item) => ({
    id: item.id,
    sourceId: item.source_id,
    sourceName: item.source_name ?? null,
    triggerType: item.trigger_type,
    status: item.status,
    startedAt: item.started_at ?? null,
    finishedAt: item.finished_at ?? null,
    nodeCount: item.node_count ?? null,
    errorCode: item.error_code ?? null,
    errorMessage: item.error_message ?? null,
    scriptLogs: mapScriptLogs(item.script_logs),
  }));

  return {
    logs,
    pagination: {
      limit: payload.pagination?.limit ?? limit,
      offset: payload.pagination?.offset ?? offset,
      total: payload.pagination?.total ?? logs.length,
      hasMore: payload.pagination?.has_more ?? false,
    },
  };
}

function mapScriptLogs(
  scriptLogs?: Array<{
    id: string;
    source_id: string;
    plugin_id: string;
    level: string;
    message: string;
    created_at: string;
  }>,
): ScriptRunLog[] {
  return (scriptLogs ?? []).map((log) => ({
    id: log.id,
    sourceId: log.source_id,
    pluginId: log.plugin_id,
    level: log.level,
    message: log.message,
    createdAt: log.created_at,
  }));
}

File: crates/app-http-server/src/handlers/logs.rs
use std::collections::HashMap;

use super::*;

const DEFAULT_LOG_LIMIT: usize = 20;
const MAX_LOG_LIMIT: usize = 200;

pub(crate) async fn list_logs_handler(
    State(state): State<ServerContext>,
    Query(query): Query<LogsQuery>,
) -> ApiResult<LogsResponse> {
    let limit = query.limit.unwrap_or(DEFAULT_LOG_LIMIT);
    if limit == 0 || limit > MAX_LOG_LIMIT {
        return Err(config_error_response("limit 必须在 1..=200 之间"));
    }
    let offset = query.offset.unwrap_or(0);

    let status_filter = query
        .status
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty());
    if let Some(status) = status_filter
        && !matches!(status, "running" | "success" | "failed")
    {
        return Err(config_error_response(
            "status 仅支持 running/success/failed",
        ));
    }
    let source_id_filter = query
        .source_id
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty());

    let refresh_repository = RefreshJobRepository::new(state.database.as_ref());
    let source_repository = SourceRepository::new(state.database.as_ref());

    let refresh_jobs = if source_id_filter.is_none() && offset == 0 && status_filter.is_some() {
        let status = status_filter.expect("status_filter 已检查为 Some");
        refresh_repository
            .list_recent_by_status(status, limit)
            .map_err(storage_error_to_response)?
    } else if source_id_filter.is_none() && offset == 0 && status_filter.is_none() {
        refresh_repository
            .list_recent(limit)
            .map_err(storage_error_to_response)?
    } else {
        refresh_repository
            .list_recent_filtered(status_filter, source_id_filter, limit, offset)
            .map_err(storage_error_to_response)?
    };
    let total = refresh_repository
        .count_filtered(status_filter, source_id_filter)
        .map_err(storage_error_to_response)?;

    let source_names = source_repository
        .list()
        .map_err(storage_error_to_response)?
        .into_iter()
        .map(|source| (source.id, source.name))
        .collect::<HashMap<_, _>>();

    let logs = refresh_jobs
        .into_iter()
        .map(|job| RefreshLogDto {
            id: job.id,
            source_id: job.source_instance_id.clone(),
            source_name: source_names.get(&job.source_instance_id).cloned(),
            trigger_type: job.trigger_type,
            status: job.status,
            started_at: job.started_at,
            finished_at: job.finished_at,
            node_count: job.node_count,
            error_code: job.error_code,
            error_message: job.error_message,
        })
        .collect();

    Ok((
        StatusCode::OK,
        Json(LogsResponse {
            logs,
            pagination: LogsPagination {
                limit,
                offset,
                total,
                has_more: offset.saturating_add(limit) < total,
            },
        }),
    ))
}
